api/types/role.go

```diff
@@ -0,0 +1,1204 @@
+	schemaDefinitions := "," + RoleSpecV3SchemaDefinitions
+	schemaTemplate := RoleSpecV3SchemaTemplate
+
+	schema := fmt.Sprintf(schemaTemplate, ``)
+	if extensionSchema != "" {
+		schema = fmt.Sprintf(schemaTemplate, ","+extensionSchema)
+	}
+
+	return fmt.Sprintf(V2SchemaTemplate, MetadataSchema, schema, schemaDefinitions)
+}
+
+// UnmarshalRole unmarshals role from JSON, sets defaults, and checks schema.
+func UnmarshalRole(data []byte, opts ...MarshalOption) (*RoleV3, error) {
+	var h ResourceHeader
+	err := json.Unmarshal(data, &h)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cfg, err := CollectOptions(opts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	switch h.Version {
+	case constants.V3:
+		var role RoleV3
+		if cfg.SkipValidation {
+			if err := utils.FastUnmarshal(data, &role); err != nil {
+				return nil, trace.BadParameter(err.Error())
+			}
+		} else {
+			if err := utils.UnmarshalWithSchema(GetRoleSchema(constants.V3, ""), &role, data); err != nil {
+				return nil, trace.BadParameter(err.Error())
+			}
+		}
+
+		if err := role.CheckAndSetDefaults(); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if cfg.ID != 0 {
+			role.SetResourceID(cfg.ID)
+		}
+		if !cfg.Expires.IsZero() {
+			role.SetExpiry(cfg.Expires)
+		}
+		return &role, nil
+	}
+
+	return nil, trace.BadParameter("role version %q is not supported", h.Version)
+}
+
+// RoleMarshaler implements marshal/unmarshal of Role implementations
+// mostly adds support for extended versions
+type RoleMarshaler interface {
+	// UnmarshalRole from binary representation
+	UnmarshalRole(bytes []byte, opts ...MarshalOption) (Role, error)
+	// MarshalRole to binary representation
+	MarshalRole(u Role, opts ...MarshalOption) ([]byte, error)
+}
+
+type teleportRoleMarshaler struct{}
+
+// UnmarshalRole unmarshals role from JSON.
+func (*teleportRoleMarshaler) UnmarshalRole(bytes []byte, opts ...MarshalOption) (Role, error) {
+	return UnmarshalRole(bytes, opts...)
+}
+
+// MarshalRole marshals role into JSON.
+func (*teleportRoleMarshaler) MarshalRole(r Role, opts ...MarshalOption) ([]byte, error) {
+	cfg, err := CollectOptions(opts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	switch role := r.(type) {
+	case *RoleV3:
+		if !cfg.PreserveResourceID {
+			// avoid modifying the original object
+			// to prevent unexpected data races
+			copy := *role
+			copy.SetResourceID(0)
+			role = &copy
+		}
+		return utils.FastMarshal(role)
+	default:
+		return nil, trace.BadParameter("unrecognized role version %T", r)
+	}
+}
+
+var roleMarshaler RoleMarshaler = &teleportRoleMarshaler{}
+
+// SetRoleMarshaler sets the global RoleMarshaler
+func SetRoleMarshaler(m RoleMarshaler) {
+	marshalerMutex.Lock()
+	defer marshalerMutex.Unlock()
+	roleMarshaler = m
+}
+
+// GetRoleMarshaler returns currently set RoleMarshaler
+func GetRoleMarshaler() RoleMarshaler {
+	marshalerMutex.Lock()
+	defer marshalerMutex.Unlock()
+	return roleMarshaler
+}
```

Comment:
@knisbet @r0mant This was used by Gravity right? Thinking about removing it if not needed by Gravity anymore.